swarm_dial.go

```diff
@@ -336,6 +318,31 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	return swarmC, nil
 }
 
+// filterKnownUndialables takes a list of multiaddrs, and removes those
+// that we definitely don't want to dial: addresses configured to be blocked,
+// IPv6 link-local addresses, addresses without a dial-capable transport,
+// and addresses that we know to be our own.
+// This is an optimization to avoid wasting time on dials that we know are going to fail.
+func (s *Swarm) filterKnownUndialables(addrs []ma.Multiaddr) []ma.Multiaddr {
+	lisAddrs, _ := s.InterfaceListenAddresses()
+	var ourAddrs []ma.Multiaddr
+	for _, addr := range lisAddrs {
+		protos := addr.Protocols()
+		// we're only sure about filtering out /ip4 and /ip6 addresses, so far
+		if len(protos) == 2 && (protos[0].Code == ma.P_IP4 || protos[0].Code == ma.P_IP6) {
```

Comment:
How about a quic addr to a filtered ip4 address? This filter won't catch it. 
I think you should drop the `len(protos) == 2` check, it's too strict.

Reply:
Maybe `len(protos) > 1`.

Reply:
No, then it'll also catch `/ip4/1.2.3.4/tcp/4001/p2p-circuit`, which should *not* be filtered.

> How about a quic addr to a filtered ip4 address?

Yeah I guess I can do that -- I'm not trying to turn this into an exhaustive list of all protocols that we have though. This filtering is only an optimization, and I'm trying to defuse it :)

Reply:
Maybe call `ToNetAddr`? But yeah, this is tricky...